pymc3/step_methods/metropolis.py

```diff
@@ -616,6 +616,145 @@ def competence(var, has_grad):
+    proposal_dist : function
+        Function that returns zero-mean deviates when parameterized with
+        S (and n). Defaults to Uniform(-S,+S).
+    scaling : scalar or array
+        Initial scale factor for epsilon. Defaults to 0.001
+    tune : str
+        Which hyperparameter to tune. Defaults to None, but can also be 'scaling' or 'lambda'.
+    tune_interval : int
+        The frequency of tuning. Defaults to 100 iterations.
+    model : PyMC Model
+        Optional model for sampling step. Defaults to None (taken from context).
+    mode :  string or `Mode` instance.
+        compilation mode passed to Theano functions
+
+    References
+    ----------
+    .. [Braak2006] Cajo C.F. ter Braak (2006).
+        Differential Evolution Markov Chain with snooker updater and fewer chains.
+        Statistics and Computing
+        `link <https://doi.org/10.1007/s11222-008-9104-9>`__
+    """
+    name = 'DEMetropolisZ'
+
+    default_blocked = True
+    generates_stats = True
+    stats_dtypes = [{
+        'accept': np.float64,
+        'accepted': np.bool,
+        'tune': np.bool,
+        'scaling': np.float64,
+        'lambda': np.float64,
+    }]
+
+    def __init__(self, vars=None, S=None, proposal_dist=None, lamb=None, scaling=0.001,
+                 tune=None, tune_interval=100, model=None, mode=None, **kwargs):
+
+        model = pm.modelcontext(model)
+
+        if vars is None:
+            vars = model.cont_vars
+        vars = pm.inputvars(vars)
+
+        if S is None:
+            S = np.ones(model.ndim)
+        
+        if proposal_dist is not None:
+            self.proposal_dist = proposal_dist(S)
+        else:
+            self.proposal_dist = UniformProposal(S)
+
+        self.scaling = np.atleast_1d(scaling).astype('d')
+        if lamb is None:
+            lamb = 2.38 / np.sqrt(2 * model.ndim)
+        self.lamb = float(lamb)
+        if not tune in {None, 'scaling', 'lambda'}:
+            raise ValueError('The parameter "tune" must be one of {None, scaling, lambda}')
+        self.tune = tune
+        self.tune_interval = tune_interval
+        self.steps_until_tune = tune_interval
+        self.accepted = 0
+
+        # cache local history for the Z-proposals
+        self._history = []
+        self._it = 0
+
+        self.mode = mode
+
+        shared = pm.make_shared_replacements(vars, model)
+        self.delta_logp = delta_logp(model.logpt, vars, shared)
+        super().__init__(vars, shared)
+
+    def astep(self, q0):
+        # same tuning scheme as DEMetropolis
+        if not self.steps_until_tune and self.tune:
+            if self.tune == 'scaling':
+                self.scaling = tune(self.scaling, self.accepted / float(self.tune_interval))
+            elif self.tune == 'lambda':
+                self.lamb = tune(self.lamb, self.accepted / float(self.tune_interval))
+            # Reset counter
+            self.steps_until_tune = self.tune_interval
+            self.accepted = 0
+
```

Comment:
Should `self._it` also be reset after tuning?

Reply:
I'm not sure yet. My first tests show that DE-MCMC has a slower initial convergence, but benefits greatly from tuning lambda. I'll run tests to answer your question.
Instead of `self._it` I can also do just `len(self._history)`.

Reply:
Resetting the history after tuning is bad. Because each chain is on its own, its equivalent to dropping the entire history.
![image](https://user-images.githubusercontent.com/5894642/72918093-5d905800-3d45-11ea-855c-94bbbd3010df.png)

Maybe if one keeps the most recent 10%, but unless the chain started far away from the typical set even that should make it worse.

Reply:
Thanks, I guess my initial feeling is that it is getting samples from tuning stage which is usually not stationary. But it is true that we are not "just" accepting a history.

Reply:
Did some further testing. It looks like a `tune_drop_fraction=0.9` does indeed improve efficiency when convergence to the typical set is slow.

![DEMetropolisZ_tuning_efficiency](https://user-images.githubusercontent.com/5894642/73015734-dca48f80-3e1c-11ea-956c-bea8eae77f6b.png)
![DEMetropolisZ_tuning_autocorrelation](https://user-images.githubusercontent.com/5894642/73015740-df9f8000-3e1c-11ea-9b74-ed504cd8ec45.png)
![DEMetropolisZ_tuning_acceptance](https://user-images.githubusercontent.com/5894642/73015743-e3330700-3e1c-11ea-9e20-8377698ed25c.png)